### PR TITLE
aie_kernels: an argmax kernel, and a stale accumulator in vector_reduce_max [MED]

### PR DESCRIPTION
**Pull request by atassis**
### Problem

Two things, one commit each, independently droppable.

**No argmax.** `aie_kernels/` has a max reduction (`aie2/reduce_max.cc`) but nothing
that returns *where* the maximum is. A caller that needs the position -- a decoder
picking a token from a logits row, any top-1 classifier -- has to read the whole
vector back to the host and scan it there.

**`vector_reduce_max` goes stale.** All five of its designs fold every tile into a
`Buffer` seeded with `initial_value=-inf`. That buffer is written at configure time,
not per run, so the second invocation of an already-compiled design starts from the
first run's maximum. Measured on NPU2, i32, 131072 elements, three inputs in one
process:

| design | run 1 (max 999999) | run 2 (max 7) | run 3 (max 42) |
|---|---|---|---|
| `single_column_designs/vector_reduce_max_chained.py` | 999999 | **999999** | **999999** |
| `single_column_designs/vector_reduce_max_shared.py` | 999999 | **999999** | **999999** |
| `single_column_designs/vector_reduce_max_memtile.py` | 999999 | **999999** | **999999** |
| `multi_column_designs/col_wise_vector_reduce_max.py` | 999999 | **999999** | **999999** |
| `multi_column_designs/row_wise_vector_reduce_max.py` | 999999 | **999999** | **999999** |

Wrong from the second run on, in all five. Each `run.lit` / `run_strix.lit` line
starts a fresh interpreter, which is why CI is green on all of them.

### Fix

**`aie_kernels/aie2/argmax.cc`** (commit 1), in the same (partial, combine) shape
that `reduce_max` + `compute_max` already use for a distributed max:
`argmax_vector*` / `argmax_scalar*` reduce a slice and `argmax_combine*` merges two
results. `int32` and `bfloat16`, matching `reduce_max`'s dtype set, and in the same
directory for the same reason -- it uses no aie2p-specific op.

- **The record.** Value and index travel together as two `int32` slots -- the value
  as itself for `int32`, widened to `float` and bit-cast for `bfloat16` -- so one
  ObjectFifo carries both and the combine step never needs the input tile.
- **Global indices.** The kernel adds an `index_offset` argument, so a core emits
  its slice's global index and the fold is order-independent, with no per-core
  fixup on the host.
- **One pass, and the tie rule falls out of it.** The vector path carries a per-lane
  running maximum and the block offset at which each lane last improved, under a
  strict `>`. Lane j only ever sees positions j, j+N, j+2N, ..., so the earliest of
  equal values survives within a lane, and the global first-occurrence index is
  `min(offset[j] + j)` over the lanes still holding the maximum -- resolved once,
  after the loop. That is `numpy.argmax`'s rule, at one read of the data. Offsets
  are int16, which bounds a call at 32767 elements: 32767 bfloat16 is 64 KB, the
  whole of a core's data memory, and a scalar tail takes any remainder so
  `tile_size` need not divide the vector width. A NaN never compares greater, so
  NaN inputs are skipped rather than returned -- `argmax_ref()` matches the kernel
  there, not numpy.

Plus the `iron.kernels.argmax` / `argmax_combine` / `argmax_ref` factories, their
`test_kernels_specs.py` rows, and `programming_examples/basic/vector_reduce_argmax`:
4 cores folding down one column, the host reading 8 bytes whatever the input length.

**The five `vector_reduce_max` designs** (commit 2) have their first tile write the
accumulator outright instead of folding into it, so nothing depends on the seed.
Three lines per core body.

### Test

`test/python/test_kernels_specs.py` + `test_kernels_memoization.py`: 330 passed.

argmax on NPU2, against `numpy.argmax`, comparing the full record (value bits and
index), 8192 elements per case:

| case | bf16 | i32 |
|---|---|---|
| 25 random seeds | 25/25 | 25/25 |
| max at index 0 / at the end | exact | exact |
| max duplicated 3x / once per core | exact | exact |
| all elements equal | exact | exact |
| strictly ascending / descending | exact | exact |

The scalar tail is covered separately, at per-core slices of 500, 300, 26 and 16
elements (16 is below the 32-lane vector width, so the tail does all of it) --
exact in every case.

All three of the new `run_strix.lit` lines pass, including the 524288-byte case that
streams 128 iterations through the memtile. Vector vs scalar on that design, 262144
bfloat16, whole dispatch, medians of 10: 0.44 ms vs 0.89 ms.

For commit 2: the table above re-run after the fix gives the correct maximum on all
three runs for all five designs, and all eight of their existing `run_strix.lit`
invocations still pass. The fix is free: interleaved before/after on `chained` and
`col_wise`, i32, 131072 elements, gives +1.7% and +2.6% against a p25-p75 spread of
0.25 and 0.43 ms.

### Scope

The kernel is device-verified on NPU2 only; it builds clean for the aie2 target but
has not been run on NPU1.

No `aie_kernels/README.md` row: #3599 is rewriting that file into a per-arch catalog,
and its table is where this entry belongs.

### CODE REVIEW

File: programming_examples/basic/vector_reduce_max/multi_column_designs/row_wise_vector_reduce_max.py [hunk 1]
```diff
@@ -122,7 +122,13 @@ def core_body(*args):
...
 
-        for _ in range_(num_iter):
+        # The first tile writes the accumulator outright rather than folding into
+        # it: a core-resident buffer keeps its value from the previous run of the
+        # same design, so a seeded accumulator makes the result depend on run order.
+        elem_in = of_in.acquire(1)
+        reduce_max_vector(elem_in, nextC_buffer, elems_per_core)
+        of_in.release(1)
+        for _ in range_(num_iter - 1):
             elem_in = of_in.acquire(1)
```
[the remark this answers is not in the document]
Comment (atassis) [reply programming_examples/basic/vector_reduce_max/multi_column_designs/row_wise_vector_reduce_max.py:136]: `_validate` already rejects this: it exits when `elems < N_MEM_ELEMS`, so `num_iter` is at least 1 and the pre-loop read always has a tile. The premise that `in1_size=512` is accepted does not hold. The comment on that check names this exact failure mode.

File: programming_examples/basic/vector_reduce_max/single_column_designs/vector_reduce_max_chained.py [hunk 2]
```diff
@@ -113,7 +113,13 @@ def start_core_body(
...
         elem_out = of_out.acquire(1)
-        for _ in range_(num_iter):
+        # The first tile writes the accumulator outright rather than folding into
+        # it: a core-resident buffer keeps its value from the previous run of the
+        # same design, so a seeded accumulator makes the result depend on run order.
+        elem_in = of_in.acquire(1)
+        reduce_max_vector(elem_in, nextC_buffer, elems_per_core)
+        of_in.release(1)
+        for _ in range_(num_iter - 1):
             elem_in = of_in.acquire(1)
```
[the remark this answers is not in the document]
Comment (atassis) [reply programming_examples/basic/vector_reduce_max/single_column_designs/vector_reduce_max_chained.py:126]: Same as the sibling designs: `_validate` exits when `elems < N_MEM_ELEMS`, so `num_iter >= 1` and the first acquire cannot outrun the fill.

File: programming_examples/basic/vector_reduce_max/single_column_designs/vector_reduce_max_memtile.py [hunk 3]
```diff
@@ -133,7 +133,13 @@ def start_core_body(
...
         elem_out = of_out.acquire(1)
-        for _ in range_(num_iter):
+        # The first tile writes the accumulator outright rather than folding into
+        # it: a core-resident buffer keeps its value from the previous run of the
+        # same design, so a seeded accumulator makes the result depend on run order.
+        elem_in = of_in.acquire(1)
+        reduce_max_vector(elem_in, nextC_buffer, elems_per_core)
+        of_in.release(1)
+        for _ in range_(num_iter - 1):
             elem_in = of_in.acquire(1)
```
[the remark this answers is not in the document]
Comment (atassis) [reply programming_examples/basic/vector_reduce_max/single_column_designs/vector_reduce_max_memtile.py:146]: Same here: `_validate` requires `elems >= N_MEM_ELEMS`, so `num_iter` is never 0.

File: programming_examples/basic/vector_reduce_max/single_column_designs/vector_reduce_max_shared.py [hunk 4]
```diff
@@ -115,7 +115,13 @@ def start_core_body(
...
     ):
         elem_out = of_out.acquire(1)
-        for _ in range_(num_iter):
+        # The first tile writes the accumulator outright rather than folding into
+        # it: a core-resident buffer keeps its value from the previous run of the
+        # same design, so a seeded accumulator makes the result depend on run order.
+        elem_in = of_in.acquire(1)
+        reduce_max_vector(elem_in, nextC_buffer, elems_per_core)
+        of_in.release(1)
+        for _ in range_(num_iter - 1):
```
[the remark this answers is not in the document]
Comment (atassis) [reply programming_examples/basic/vector_reduce_max/single_column_designs/vector_reduce_max_shared.py:127]: `in1_size=512` for i32 gives 128 elements, and `_validate` exits on `elems < N_MEM_ELEMS` (2048) before the design is built, so this path is not reachable from the CLI.
